### ISSUE TITLE
String format %c pretty incorrect

### ISSUE THREAD

**Issue by chrisseaton** [oracle/truffleruby#2369]
```ruby
puts RUBY_DESCRIPTION
p sprintf('%c', 0xf).bytes
p sprintf('%c', 0xff).bytes
p sprintf('%c', 0xffff).bytes
p sprintf('%c', 0xfffffff).bytes
```

```
ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-darwin20]
[15]
[195, 191]
[239, 191, 191]
Traceback (most recent call last):
	1: from test.rb:5:in `<main>'
test.rb:5:in `sprintf': invalid character (ArgumentError)
```

```
truffleruby 21.2.0-dev-55ec022f, like ruby 2.7.3, GraalVM CE JVM [x86_64-darwin]
[15]
[255]
[255]
[255]
```

Found while working on #2368.

**Comment by eregon** [comment]
We also noticed this one when running https://github.com/ruby/ruby/blob/master/benchmark/app_aobench.rb (GR-13417).
PR welcome, it seems a bug in format nodes.

My notes from then:
This turns out to be a bug in Kernel#sprintf. Replacing with `IO#write(byte.chr(Encoding::BINARY))` produces the correct image.

The logic there seems to mostly ignore encodings, which is obviously wrong.

There is already a failing spec showing the problem:

"Kernel#sprintf other formats c supports Unicode characters" which can be run with:
```
$ jt test spec/ruby/core/kernel/sprintf_spec.rb
```
I started to work on it but gave up because this logic is fundamentally problematic as it uses Java strings for formatting, losing encoding information, instead of using byte[] or Ruby String methods.

I push my rebased WIP at https://github.com/oracle/truffleruby/compare/master...eregon:fix-format-c
